File: 05_Abstraction/Abstract_Classes/Main.cpp
#include <iostream>
using namespace std;

class Animal {
public:
    virtual void sound() = 0;

    void sleep() {
        cout << "This animal sleeps." << endl;
    }
};

class Dog : public Animal {
public:
    void sound() override {
        cout << "The dog barks." << endl;
    }
};

class Cat : public Animal {
public:
    void sound() override {
        cout << "The cat meows." << endl;
    }
};

int main() {
    Animal* dog = new Dog();
    Animal* cat = new Cat();

    dog->sound(); 
    cat->sound(); 
    dog->sleep(); 
    cat->sleep(); 

    delete dog;
    delete cat;
    return 0;
}

File: 05_Abstraction/Abstract_Classes/Main.java
abstract class Animal {
    abstract void sound();

    void sleep() {
        System.out.println("This animal sleeps.");
    }
}

class Dog extends Animal {
    void sound() {
        System.out.println("The dog barks.");
    }
}

class Cat extends Animal {
    void sound() {
        System.out.println("The cat meows.");
    }
}

public class Main {
    public static void main(String[] args) {
        Animal dog = new Dog();
        Animal cat = new Cat();

        dog.sound(); 
        cat.sound(); 
        dog.sleep(); 
        cat.sleep(); 
    }
}
